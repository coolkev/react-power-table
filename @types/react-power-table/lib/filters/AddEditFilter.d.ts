/// <reference types="react" />
import * as React from 'react';
import * as PowerTable from "./definitions/FilterDefinition";
export interface AddEditFilterProps {
    filter: PowerTable.FilterDefinition<any>;
    initialOperation: PowerTable.OperationDefinition<any>;
    initialValue?: any;
    onApplyFilter: (filter: PowerTable.AppliedFilter<any>) => void;
}
export interface AddEditFilterState {
    operation: PowerTable.OperationDefinition<any>;
    value: any;
}
export declare class AddEditFilter extends React.PureComponent<AddEditFilterProps, AddEditFilterState> {
    constructor(props: AddEditFilterProps);
    private selectedOperationChange(e);
    private handleFilterValueChange(value);
    private applyFilter();
    render(): JSX.Element;
}
