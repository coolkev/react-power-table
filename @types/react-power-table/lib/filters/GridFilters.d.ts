/// <reference types="react" />
import * as React from 'react';
import * as PowerTable from "./definitions/FilterDefinition";
export interface GridFiltersProps {
    availableFilters: PowerTable.AvailableFiltersMap;
    appliedFilters: PowerTable.AppliedFilterType[];
    onFiltersChange: (newFilters: PowerTable.AppliedFilterType[]) => void;
}
export interface GridFiltersState {
    addingFilter?: boolean;
    editingFilter?: PowerTable.AppliedFilterType;
}
export interface GridAppliedFiltersProps {
    availableFilters: PowerTable.AvailableFiltersMap;
    appliedFilters: PowerTable.AppliedFilterType[];
    removeFilter: (filter: PowerTable.AppliedFilterType) => void;
    editFilter: (filter: PowerTable.AppliedFilterType) => void;
    onOptionLoaded: () => void;
}
export declare class GridFilters extends React.Component<GridFiltersProps, GridFiltersState> {
    constructor(props: GridFiltersProps);
    private showAddFilter();
    private hideAddFilter();
    private applyNewfilter(filter);
    private removeFilter(filter);
    private editFilter(filter);
    private cancelEditFilter();
    private applyEditFilter(filter);
    render(): JSX.Element;
}
