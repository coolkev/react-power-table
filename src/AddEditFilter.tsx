import * as React from 'react';
import * as Button from 'react-bootstrap/lib/Button';
import * as Radio from 'react-bootstrap/lib/Radio';
import * as filters from './filters/FilterDefinition';
import { debuglog } from './utils';

/**
 * @internal
 */
export interface AddEditFilterProps {
    filter: filters.FilterDefinition<any>;
    initialOperation: filters.OperationDefinition<any>;
    initialValue?: any;
    onApplyFilter: (filter: filters.AppliedFilter<any>) => void;
}

/**
 * @internal
 */
export interface AddEditFilterState {

    operationKey: string;
    value: any;

}

/**
 * @internal
 */
export class AddEditFilter extends React.PureComponent<AddEditFilterProps, AddEditFilterState> {

    constructor(props: AddEditFilterProps) {
        super(props);

        const value = props.initialValue || props.filter.defaultValue || '';

        this.state = { value, operationKey: props.initialOperation.key };

        this.selectedOperationChange = this.selectedOperationChange.bind(this);
        this.handleFilterValueChange = this.handleFilterValueChange.bind(this);
        this.applyFilter = this.applyFilter.bind(this);

    }

    private selectedOperationChange(e: React.FormEvent<Radio & HTMLInputElement>) {

        const prevOperationKey = this.state.operationKey;
        const operationKey = e.currentTarget.value;
        //const operation = this.props.filter.operations[operationKey];

        const newState: AddEditFilterState = { operationKey, value: this.state.value };

        this.props.filter.onOperationChange(prevOperationKey, newState);
        // if (prevOperationKey === 'between' && operationKey !== prevOperationKey && typeof this.state.value === 'string') {
        //     newState.value = this.state.value.split(' ')[0];
        // }
        this.setState(newState);
    }

    private handleFilterValueChange(value: string) {
        debuglog('AddEditFilter.filterValueChange', value);
        this.setState({
            value,
        });
    }

    private applyFilter() {
        const operation = this.props.filter.operations[this.state.operationKey];

        this.props.onApplyFilter({ filter: this.props.filter, operation, value: this.state.value });
    }
    render() {
        debuglog('AddEditFilter.render', this.props);

        const filter = this.props.filter;

        const { value, operationKey } = this.state;

        const opComponents = Object.keys(filter.operations).map((opKey) => {
            const op = filter.operations[opKey];
            let children: React.ReactChild;
            if (opKey === this.state.operationKey) {

                const SelectedFilterComponent = op.filterComponent || filter.filterComponent;

                if (SelectedFilterComponent) {
                    children = <div style={{ marginLeft: 20 }}><SelectedFilterComponent filter={filter} operation={op} value={value} onValueChange={this.handleFilterValueChange} onEnterKeyPress={this.applyFilter} /></div>;
                }
            }

            return <FilterOperation key={opKey} operation={op} filter={filter} selected={opKey === operationKey} onChange={this.selectedOperationChange}>{children}</FilterOperation>;
        });

        return (
            <div>
                <div style={{ margin: '10px 0' }}><b>{filter.displayName}</b>
                </div>

                {opComponents.map((opComponent) => opComponent)}

                <div style={{ marginTop: 20 }}>
                    <Button bsStyle="primary" bsSize="sm" onClick={this.applyFilter}>Apply Filter</Button>
                </div>

            </div>);
    }
}

/**
 * @internal
 */
export interface FilterOperationProps extends React.Props<any> {
    operation: filters.OperationDefinition;
    filter: filters.FilterDefinition;

    selected: boolean;
    onChange: React.EventHandler<React.FormEvent<Radio & HTMLInputElement>>;
}

/**
 * @internal
 */
export const FilterOperation = (props: FilterOperationProps) => {

    const { operation, filter, onChange, selected, children } = props;

    return (
        <div key={operation.key}>
            <Radio checked={selected} onChange={onChange} value={operation.key}>
                {filter.radioButtonLabel ? filter.radioButtonLabel({ filter, operation }) : operation.displayName}

            </Radio>
            {children}
        </div>
    );

};
