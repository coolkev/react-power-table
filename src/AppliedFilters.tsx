import * as React from 'react';
import { ActionLink } from './components/ActionLink';
import * as filters from './filters';

export interface AppliedFiltersProps {
    appliedFilters: ReadonlyArray<filters.AppliedFilter>;
    removeFilter: (filter: filters.AppliedFilter) => void;
    editFilter: (filter: filters.AppliedFilter) => void;

}

export const AppliedFilters = (props: AppliedFiltersProps) => {
    return (
        <div className="applied-filters small">
            {props.appliedFilters.map((appliedFilter) => <AppliedFilter appliedFilter={appliedFilter} key={appliedFilter.key} onEditFilter={props.editFilter} onRemoveFilter={props.removeFilter} />)}
        </div>
    );
};

export interface AppliedFilterProps {
    appliedFilter: filters.AppliedFilter;
    onRemoveFilter: (filter: filters.AppliedFilter) => void;
    onEditFilter: (filter: filters.AppliedFilter) => void;

}

const defaultAppliedFilterLabelComponent = (props: filters.AppliedFilter<any>) => {
    const appliedLabel = props.operation.appliedLabel || props.filter.appliedLabel || filters.FilterDefinition.defaultAppliedFilterLabel;
    return <span>{appliedLabel(props)}</span>;
};

export const AppliedFilter = ({ appliedFilter, onRemoveFilter, onEditFilter }: AppliedFilterProps) => {

    const handleRemoveFilter = React.useCallback((e: React.SyntheticEvent<any>) => {
        e.preventDefault();
        onRemoveFilter(appliedFilter);
    }, [appliedFilter, onRemoveFilter]);

    const handleEditFilter = React.useCallback(() => {

        onEditFilter(appliedFilter);

    }, [appliedFilter, onEditFilter]);

    const AppliedLabelComponent = appliedFilter.operation.appliedLabelComponent || appliedFilter.filter.appliedLabelComponent || defaultAppliedFilterLabelComponent;

    return (
        <div className="applied-filter well well-sm" style={{ marginBottom: 10 }} key={appliedFilter.filter.fieldName}>
            <button type="button" className="close" aria-label="Remove" onClick={handleRemoveFilter}><span aria-hidden="true">Ã—</span></button>
            <ActionLink onClick={handleEditFilter}><AppliedLabelComponent {...appliedFilter} /></ActionLink>
        </div>
    );

};

