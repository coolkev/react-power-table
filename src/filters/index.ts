import * as Select from 'react-select';
import { BooleanFilter } from './BooleanFilter';
import { DateFilter } from './DateFilter';
import { DecimalFilter } from './DecimalFilter';
import { FilterDefinition, FilterDefinitionOptionsOrFieldName, OperationDefinition } from './FilterDefinition';
import { IntFilter } from './IntFilter';
import { ListFilter } from './ListFilter';
import { RemoteListFilter, RemoteListOptionProvider } from './RemoteListFilter';
import { StringFilter } from './StringFilter';
import { TimeFilter } from './TimeFilter';

export * from './FilterDefinition';

//type Types = 'string' | 'boolean' | 'date' | 'decimal' | 'int';

const types = {
    string: StringFilter,
    boolean: BooleanFilter,
    date: DateFilter,
    decimal: DecimalFilter,
    int: IntFilter,
    list: ListFilter,
    remotelist: RemoteListFilter,
    timespan: TimeFilter,
};
type types = typeof types & { [key: string]: FilterDefinition<any> };

export function getFilterDefinition(type: 'string', options: FilterDefinitionOptionsOrFieldName): StringFilter;
export function getFilterDefinition(type: 'boolean', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): BooleanFilter & { operations: { 'notnull': OperationDefinition<boolean>, 'null': OperationDefinition<boolean> } };
export function getFilterDefinition(type: 'boolean', options: FilterDefinitionOptionsOrFieldName): BooleanFilter;
export function getFilterDefinition(type: 'date', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): DateFilter & { operations: { 'notnull': OperationDefinition<string>, 'null': OperationDefinition<string> } };
export function getFilterDefinition(type: 'date', options: FilterDefinitionOptionsOrFieldName): DateFilter;
export function getFilterDefinition(type: 'time', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): TimeFilter & { operations: { 'notnull': OperationDefinition<string>, 'null': OperationDefinition<string> } };
export function getFilterDefinition(type: 'time', options: FilterDefinitionOptionsOrFieldName): TimeFilter;
export function getFilterDefinition(type: 'decimal', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): DecimalFilter & { operations: { 'notnull': OperationDefinition<number>, 'null': OperationDefinition<number> } };
export function getFilterDefinition(type: 'decimal', options: FilterDefinitionOptionsOrFieldName): DecimalFilter;
export function getFilterDefinition(type: 'int', options: FilterDefinitionOptionsOrFieldName & { canBeNull: true }): IntFilter & { operations: { 'notnull': OperationDefinition<number>, 'null': OperationDefinition<number> } };
export function getFilterDefinition(type: 'int', options: FilterDefinitionOptionsOrFieldName): IntFilter;
export function getFilterDefinition(type: 'list', options: FilterDefinitionOptionsOrFieldName, items: ReadonlyArray<Select.Option>): ListFilter;
export function getFilterDefinition(type: 'remotelist', options: FilterDefinitionOptionsOrFieldName, optionProvider: RemoteListOptionProvider): RemoteListFilter;

export function getFilterDefinition(type: string, options: FilterDefinitionOptionsOrFieldName, items?: ReadonlyArray<Select.Option> | RemoteListOptionProvider) {

    return new types[type](options, items);
}

export const DataTypes = types;
