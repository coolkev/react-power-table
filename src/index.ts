export * from './ReactPowerTable';
export * from './Paging';
export * from './Sorting';
export * from './GridFilters';
export * from './filters/';
export * from './Column';
export * from './utils';

import { PowerTableProps } from './ReactPowerTable';
export type ReactPowerTable<TRow = {}, TExtraProps = {}> = React.ComponentClass<PowerTableProps<TRow, TExtraProps>>;
