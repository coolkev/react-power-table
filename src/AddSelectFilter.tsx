import * as React from 'react';
import * as FormControl from 'react-bootstrap/lib/FormControl';
import { AddEditFilter } from './AddEditFilter';
import { BackLink } from './components/BackLink';
import * as PowerTable from './filters/FilterDefinition';
import { objectMapToArray } from './utils';

/**
 * @internal
 */
export interface AddFilterProps {
    cancelAddFilter: () => void;

    availableFilters: PowerTable.FilterDefinition[] | { [key: string]: PowerTable.FilterDefinition };

    //availableFilters: { [key: string]: PowerTable.FilterDefinition };
    appliedFilters: PowerTable.AppliedFilter[];
    onApplyFilter: (filter: PowerTable.AppliedFilter) => void;

}

/**
 * @internal
 */
export interface AddFilterState {
    searchText: string;
    selectedFilterKey?: string;
}

/**
 * @internal
 */
export class AddSelectFilter extends React.PureComponent<AddFilterProps, AddFilterState> {

    constructor(props: AddFilterProps) {
        super(props);
        this.state = { searchText: '' };
        this.onSearchTextChanged = this.onSearchTextChanged.bind(this);
        this.backToPrev = this.backToPrev.bind(this);
        this.newFilterSelected = this.newFilterSelected.bind(this);

    }

    private onSearchTextChanged(e: React.FormEvent<FormControl & HTMLInputElement>) {
        this.setState({
            searchText: e.currentTarget.value,
        });

    }

    private newFilterSelected(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        this.setState({
            selectedFilterKey: e.currentTarget.dataset.fieldName,
        });
    }

    private backToPrev(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        if (this.state.selectedFilterKey) {
            this.setState({ selectedFilterKey: null });
        } else {
            this.props.cancelAddFilter();
        }
    }
    render() {

        const { availableFilters, appliedFilters } = this.props;

        const { searchText, selectedFilterKey } = this.state;

        if (selectedFilterKey) {

            const filter = Array.isArray(availableFilters) ? availableFilters.find((f) => f.fieldName === selectedFilterKey) : availableFilters[selectedFilterKey];

            const initialOperation = objectMapToArray(filter.operations)[0];

            return (
                <div>
                    <BackLink onClick={this.backToPrev} />

                    <div>

                        <AddEditFilter filter={filter} initialOperation={initialOperation} onApplyFilter={this.props.onApplyFilter} />

                    </div>
                </div>
            );

        }

        const filters = objectMapToArray(availableFilters);

        const unusedFilters = filters.filter((m) => appliedFilters.every((c) => c.filter.fieldName !== m.fieldName));

        const regex = new RegExp(searchText, 'i');
        const showFilters = searchText ? unusedFilters.filter((m) => m.displayName.match(regex)) : unusedFilters;

        // const onSelectFilter = (e: React.) => {
        //     e.preventDefault();
        //     this.newFilterSelected(m.fieldName);
        // };

        return (
            <div className="flex-column">
                <BackLink onClick={this.backToPrev} />

                <div><FormControl placeholder="Filter by" value={searchText} onChange={this.onSearchTextChanged} autoFocus /></div>

                <div className="small flex-column">
                    <div style={{ margin: '10px 0' }}><b>Available Filters</b></div>
                    <div className="available-filters">
                        <div className="list-group">
                            {showFilters.map((m) => <a href="#" onClick={this.newFilterSelected} data-fieldName={m.fieldName} className="list-group-item" key={m.fieldName}>{m.displayName}</a>)}
                        </div>
                    </div>

                </div>

            </div>
        );
    }
}
