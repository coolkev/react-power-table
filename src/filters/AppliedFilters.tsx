import * as React from 'react';
import { AddEditFilter } from './AddEditFilter'
import FormControl from 'react-bootstrap/lib/FormControl';
import * as PowerTable from "./definitions/FilterDefinition";
import { BackLink } from "./BackLink";
import { AddSelectFilter } from "./AddSelectFilter";


/**
  * @internal
  */
export interface AppliedFiltersProps {
    availableFilters: PowerTable.AvailableFiltersMap;
    appliedFilters: PowerTable.AppliedFilterType[];
    removeFilter: (filter: PowerTable.AppliedFilterType) => void;
    editFilter: (filter: PowerTable.AppliedFilterType) => void;
    onOptionLoaded: () => void;
}



/**
  * @internal
  */
export const AppliedFilters = (props: AppliedFiltersProps) => {

    return <div className="small">
        {props.appliedFilters.map(appliedFilter => {

            let AppliedLabelComponent = appliedFilter.operation.appliedLabelComponent || appliedFilter.filter.appliedLabelComponent;

            if (!AppliedLabelComponent) {
                const appliedLabel = appliedFilter.operation.appliedLabel || appliedFilter.filter.appliedLabel || PowerTable.FilterDefinition.defaultAppliedFilterLabel;
                AppliedLabelComponent = (props: PowerTable.AppliedFilter<any>) => <span>{appliedLabel(props)}</span>;
            }
            //const formatResult = formatFunc(appliedFilter);

            {/*let displayText: string;
            if (typeof (formatResult) == 'string') {
                displayText = formatResult
            } else {
                formatResult.then(() => props.onOptionLoaded());
                displayText = 'Loading...';
            }*/}
            return <div className="well well-sm" style={{ marginBottom: 10 }} key={appliedFilter.filter.fieldName}>
                <button type="button" className="close" aria-label="Remove" onClick={() => props.removeFilter(appliedFilter)}><span aria-hidden="true">Ã—</span></button>
                <a href="#" onClick={e => { e.preventDefault(); props.editFilter(appliedFilter); }}><AppliedLabelComponent {...appliedFilter} /></a>
            </div>;
        }
        )}



    </div>

};
