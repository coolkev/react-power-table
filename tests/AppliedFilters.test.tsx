import * as React from 'react';
import { GridFilters, DataTypes, createKeyedMap, AppliedFilter } from '../src/';
import { defaultColumns, sampledata, President } from "./shared";
import { mount } from 'enzyme';
import { AppliedFilters } from "../src/filters/AppliedFilters";

//const columns = defaultColumns;
const rows = sampledata.slice(0, 25);

const availableFilters = [
    new DataTypes.int('number'),
    new DataTypes.string('president'),
    new DataTypes.int('birth_year'),
    new DataTypes.int({ fieldName: 'death_year', canBeNull: true }),
    new DataTypes.date({ fieldName: 'took_office' }),
    new DataTypes.date({ fieldName: 'left_office', canBeNull: true }),
    new DataTypes.boolean({ fieldName: 'assasinated', displayName: 'was assasinated' }),

];
const availableFiltersMap = createKeyedMap(availableFilters, m => m.fieldName);
describe('AppliedFilters tests',
    () => {



        test('none applied', () => {

            const c = <AppliedFilters availableFilters={availableFiltersMap} appliedFilters={[]} editFilter={() => { }} removeFilter={() => { }} />;

            const component = mount(c);

            expect(component.render()).toMatchSnapshot();
        });



        test('one applied', () => {

            const applied: AppliedFilter<any> = {
                filter: availableFiltersMap.president,
                operation: availableFiltersMap.president.operations.contains,
                value: 'george'
            };

            const c = <AppliedFilters availableFilters={availableFiltersMap} appliedFilters={[applied]} editFilter={() => { }} removeFilter={() => { }} />;

            const component = mount(c);

            expect(component.render()).toMatchSnapshot();
        });

    });
