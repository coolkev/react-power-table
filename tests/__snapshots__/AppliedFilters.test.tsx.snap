// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppliedFilters tests none applied 1`] = `
<div
  class="small"
/>
`;

exports[`AppliedFilters tests one applied 1`] = `
<div
  class="small"
>
  <div
    class="well well-sm"
    style="margin-bottom: 10px;"
  >
    <button
      aria-label="Remove"
      class="close"
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
    </button>
    <a
      href="#"
    >
      <span>
        president contains "george"
      </span>
    </a>
  </div>
</div>
`;
